## CoNLL-2003 Example for Text Extensions for Pandas
### Part 2

To run this notebook, you will need to obtain a copy of the CoNLL-2003 data set's corpus.
Drop the corpus's files into the following locations:
* conll_03/eng.testa
* conll_03/eng.testb
* conll_03/eng.train

If you are unfamiliar with the basics of Text Extensions for Pandas, we recommend you 
start with Part 1 of this example.

### Introduction

In Part 1 of this demo, we showed how to use Text Extensions for Pandas to examine the 
overall result quality of one entrant in the CoNLL-2003 Shared Task, as well as how to
identify and examine the documents from the validation set where that entry had the 
most errors.

In Part 2, we'll perform a broader analysis that goes across all the contents entries
and come to some deeper and more surprising conclusions.

In [1]:
# INITIALIZATION BOILERPLATE

# The Jupyter kernel for this notebook usually starts up inside the notebooks
# directory, but the text_extensions_for_pandas package code is in the parent
# directory. Add that parent directory to the front of the Python include path.
import sys
if (sys.path[0] != ".."):
    sys.path[0] = ".."
    
# Libraries
import numpy as np
import pandas as pd

# And of course we need the text_extensions_for_pandas library itself.
import text_extensions_for_pandas as tp

In [2]:
# Load up the same gold standard data we used in Part 1.
gold_standard = tp.conll_2003_to_dataframes("../conll_03/eng.testa")
gold_standard_spans = [tp.iob_to_spans(df) for df in gold_standard]

In [3]:
# Load up the results from all 16 teams at once.
teams = ["bender", "carrerasa", "carrerasb", "chieu", "curran",
         "demeulder", "florian", "hammerton", "hendrickx",
         "klein", "mayfield", "mccallum", "munro", "whitelaw",
         "wu", "zhang"]

# Read all the output files into one dataframe per <document, team> pair.
outputs = { 
    t: tp.conll_2003_output_to_dataframes(
        gold_standard, f"../resources/conll_03/ner/results/{t}/eng.testb")
    for t in teams
}  # Type: Dict[str, List[pd.DataFrame]]

# As an example of what we just loaded, show the token metadata for the 
# "mayfield" team's model's output on document 3.
outputs["mayfield"][3]

char_span                token_span ent_iob ent_type  \
0       [0, 10): '-DOCSTART-'     [0, 10): '-DOCSTART-'       O     None   
1       [11, 20): 'FREESTYLE'     [11, 20): 'FREESTYLE'       O     None   
2    [21, 33): 'SKIING-WORLD'  [21, 33): 'SKIING-WORLD'       O     None   
3             [34, 37): 'CUP'           [34, 37): 'CUP'       B     MISC   
4           [38, 43): 'MOGUL'         [38, 43): 'MOGUL'       O     None   
..                        ...                       ...     ...      ...   
161      [803, 809): 'Allais'      [803, 809): 'Allais'       I      PER   
162           [809, 810): '('           [809, 810): '('       O     None   
163      [811, 817): 'France'      [811, 817): 'France'       B      LOC   
164           [817, 818): ')'           [817, 818): ')'       O     None   
165       [819, 824): '21.58'       [819, 824): '21.58'       O     None   

                                              sentence  
0                                [0, 10): '-DOCSTART-'  
1    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
2    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
3    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
4    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
..                                                 ...  
161    [791, 824): '10. Katleen Allais( France) 21.58'  
162    [791, 824): '10. Katleen Allais( France) 21.58'  
163    [791, 824): '10. Katleen Allais( France) 21.58'  
164    [791, 824): '10. Katleen Allais( France) 21.58'  
165    [791, 824): '10. Katleen Allais( France) 21.58'  

[166 rows x 5 columns]

In [4]:
# Convert results from IOB2 tags to spans across all teams and documents
# See https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging) for details on IOB2 format.
output_spans = {
    t: [tp.iob_to_spans(df) for df in outputs[t]] for t in teams
}  # Type: Dict[str, List[pd.DataFrame]]

# As an example, show the first 10 spans that the "florian" team's model
# found on document 2.
output_spans["florian"][2].head(10)

token_span ent_type
0                 [35, 40): 'JAPAN'      LOC
1                 [50, 55): 'SYRIA'      LOC
2                [57, 63): 'AL-AIN'      LOC
3  [65, 85): 'United Arab Emirates'      LOC
4               [144, 149): 'Japan'      LOC
5           [169, 178): 'Asian Cup'     MISC
6               [192, 197): 'Syria'      LOC
7       [209, 222): 'Takuya Takagi'      PER
8         [297, 308): 'Salem Bitar'      PER
9              [403, 409): 'Syrian'     MISC

In [5]:
# Use Pandas merge to find what spans match up exactly for each team's
# results.
# Unlike Part 1, we perform the join across all entity types, looking for
# matches of both the extracted span *and* the entity type label.
def make_stats_df(gold_dfs, output_dfs):
    num_true_positives = [len(gold_dfs[i].merge(output_dfs[i]).index)
                          for i in range(len(gold_dfs))]
    num_extracted = [len(df.index) for df in output_dfs]
    num_entities = [len(df.index) for df in gold_dfs]
    doc_num = np.arange(len(gold_dfs))

    stats_by_doc = pd.DataFrame({
        "doc_num": doc_num,
        "num_true_positives": num_true_positives,
        "num_extracted": num_extracted,
        "num_entities": num_entities
    })
    stats_by_doc["precision"] = stats_by_doc["num_true_positives"] / stats_by_doc["num_extracted"]
    stats_by_doc["recall"] = stats_by_doc["num_true_positives"] / stats_by_doc["num_entities"]
    stats_by_doc["F1"] = 2.0 * (stats_by_doc["precision"] * stats_by_doc["recall"]) / (stats_by_doc["precision"] + stats_by_doc["recall"])
    return stats_by_doc

stats = {t: make_stats_df(gold_standard_spans, output_spans[t]) for t in teams}

# Show the result quality statistics by document for the "carrerasa" team
stats["carrerasa"]

doc_num  num_true_positives  num_extracted  num_entities  precision  \
0          0                  42             48            45   0.875000   
1          1                  43             44            44   0.977273   
2          2                  51             52            54   0.980769   
3          3                  43             44            44   0.977273   
4          4                  14             19            19   0.736842   
..       ...                 ...            ...           ...        ...   
226      226                   7              7             7   1.000000   
227      227                  19             22            21   0.863636   
228      228                  22             27            27   0.814815   
229      229                  26             27            27   0.962963   
230      230                  22             27            28   0.814815   

       recall        F1  
0    0.933333  0.903226  
1    0.977273  0.977273  
2    0.944444  0.962264  
3    0.977273  0.977273  
4    0.736842  0.736842  
..        ...       ...  
226  1.000000  1.000000  
227  0.904762  0.883721  
228  0.814815  0.814815  
229  0.962963  0.962963  
230  0.785714  0.800000  

[231 rows x 7 columns]

In [6]:
# F1 for document 4 is looking a bit low. Is that just fluke, or is it
# part of a larger trend? 
# In Part 1, we showed how to drill down to and examine "problem" documents.
# Since we have all this additional data, let's try a broader, more 
# quantitative approach. We'll start by building up some more fine-grained 
# data about congruence between the gold standard and the model outputs.
# Pandas' outer join will tell us what entities showed up just in the gold
# standard, just in the model output, or in both sets.
# For starters, let's do this just for the "carrerasa" team and document 4.
gold_standard_spans[4].merge(output_spans["carrerasa"][4], how="outer", indicator=True).sort_values("token_span")

token_span ent_type      _merge
0              [19, 28): 'ASIAN CUP'     MISC        both
1                 [46, 52): 'AL-AIN'      LOC        both
2   [54, 74): 'United Arab Emirates'      LOC        both
3             [97, 106): 'Asian Cup'     MISC        both
4                [141, 146): 'Japan'      LOC   left_only
19               [141, 146): 'Japan'      ORG  right_only
5                [149, 154): 'Syria'      LOC   left_only
20               [149, 154): 'Syria'      ORG  right_only
6                [181, 186): 'Japan'      LOC        both
7         [188, 200): 'Hassan Abbas'      PER        both
8        [214, 227): 'Takuya Takagi'      PER        both
9                [232, 237): 'Syria'      LOC        both
10      [239, 253): 'Nader Jokhadar'      PER        both
21               [276, 281): 'China'      ORG  right_only
11               [276, 281): 'China'      LOC   left_only
12          [284, 294): 'Uzbekistan'      LOC   left_only
22          [284, 294): 'Uzbekistan'      ORG  right_only
13       [321, 334): 'Shkvyrin Igor'      PER        both
14      [339, 353): 'Shatskikh Oleg'      PER        both
15          [462, 472): 'Uzbekistan'      LOC        both
16               [487, 492): 'Japan'      LOC   left_only
23               [487, 492): 'Japan'      ORG  right_only
17               [507, 512): 'Syria'      LOC        both
18               [527, 532): 'China'      LOC        both

In [7]:
# Repeat the analysis from the previous cell across all teams and documents.
# That is, perform an outer join between the gold standard spans dataframe
# for each document and the corresponding dataframe from each team.
def merge_span_sets(team):
    result = []
    for i in range(len(gold_standard_spans)):
        merged = gold_standard_spans[i].merge(output_spans[team][i], how="outer", indicator=True)
        merged["gold"] = merged["_merge"].isin(("both", "left_only"))
        merged[team] = merged["_merge"].isin(("both", "right_only"))
        result.append(merged[["token_span", "ent_type", "gold", team]])
    return result

span_flags = {t: merge_span_sets(t) for t in teams}  # Type: Dict[str, List[pd.DataFrame]]

In [8]:
# Now we have indicator variables for every extracted span, telling whether 
# it was in the gold standard data set and/or in each of the team's results.
# For example, here are the first 5 spans for document 2 in the "carrerasa"
# team's results:
span_flags["carrerasa"][2].head(50)

token_span ent_type  gold  carrerasa
0                      [35, 40): 'JAPAN'      LOC  True       True
1                      [50, 55): 'SYRIA'      LOC  True       True
2                     [57, 63): 'AL-AIN'      LOC  True       True
3       [65, 85): 'United Arab Emirates'      LOC  True       True
4                    [144, 149): 'Japan'      LOC  True       True
5                [169, 178): 'Asian Cup'     MISC  True       True
6                    [192, 197): 'Syria'      LOC  True       True
7            [209, 222): 'Takuya Takagi'      PER  True       True
8              [297, 308): 'Salem Bitar'      PER  True       True
9                   [403, 409): 'Syrian'     MISC  True       True
10            [454, 466): 'Hassan Abbas'      PER  True       True
11                   [583, 588): 'Bitar'      PER  True       True
12                   [598, 603): 'Syria'      LOC  True       True
13          [678, 692): 'Nader Jokhadar'      PER  True       True
14              [726, 736): 'Ammar Awad'      PER  True       True
15       [766, 783): 'Kenichi Shimokawa'      PER  True       True
16                   [793, 798): 'Japan'      LOC  True       True
17                  [822, 828): 'Syrian'     MISC  True       True
18                   [916, 921): 'Bitar'      PER  True       True
19         [962, 977): 'Kazuyoshi Miura'      PER  True       True
20                [1007, 1013): 'Takagi'      PER  True       True
21                 [1030, 1035): 'Bitar'      PER  True      False
22                 [1058, 1063): 'Miura'      PER  True       True
23                 [1124, 1129): 'Japan'      LOC  True       True
24                 [1167, 1172): 'Bitar'      PER  True       True
25            [1233, 1243): 'Naoki Soma'      PER  True       True
26                 [1277, 1282): 'Japan'      LOC  True       True
27     [1288, 1305): 'Kenichi Shimokawa'      PER  True       True
28  [1310, 1330): 'Hiroshige Yanagimoto'      PER  True       True
29            [1335, 1345): 'Naoki Soma'      PER  True       True
30          [1350, 1362): 'Masami Ihara'      PER  True       True
31           [1367, 1378): 'Norio Omura'      PER  True       True
32    [1383, 1401): 'Motohiro Yamaguchi'      PER  True       True
33      [1406, 1422): 'Masakiyo Maezono'      PER  True       True
34          [1427, 1439): 'Yasuto Honda'      PER  True       True
35         [1448, 1461): 'Takuya Takagi'      PER  True       True
36        [1467, 1481): 'Hiroshi Nanami'      PER  True       True
37       [1487, 1502): 'Kazuyoshi Miura'      PER  True       True
38     [1508, 1525): 'Hiroaki Morishima'      PER  True       True
39        [1531, 1545): 'Masayuki Okano'      PER  True       True
40                 [1551, 1556): 'Syria'      LOC  True       True
41           [1562, 1573): 'Salem Bitar'      PER  True       True
42          [1578, 1590): 'Bachar Srour'      PER  True       True
43          [1595, 1607): 'Hassan Abbas'      PER  True       True
44          [1612, 1624): 'Tarek Jabban'      PER  True       True
45            [1629, 1639): 'Ammar Awad'      PER  True       True
46           [1644, 1655): 'Louay Taleb'      PER  True      False
47       [1664, 1679): 'Nihad al-Boushi'      PER  True      False
48        [1685, 1699): 'Mohammed Afash'      PER  True       True
49               [1705, 1712): 'Ali Dib'      PER  True       True

In [9]:
# Do an n-way merge of all those indicator variables across documents.
# This operation produces a single summary dataframe per document.
indicators = []  # Type: List[pd.DataFrame]
for i in range(len(gold_standard_spans)):
    result = gold_standard_spans[i]
    for t in teams:
        result = result.merge(span_flags[t][i], how="outer")
    indicators.append(result.fillna(False))

In [10]:
# Now we have a vector of indicator variables for every span extracted 
# from every document across all the model outputs and the gold standard.
# For example, let's show the results for document 10:
indicators[10]

token_span ent_type   gold  bender  carrerasa  \
0        [11, 22): 'RUGBY UNION'      ORG   True    True       True   
1             [24, 30): 'LITTLE'      PER   True    True       True   
2            [39, 46): 'CAMPESE'      PER   True   False      False   
3      [57, 70): 'Robert Kitson'      PER   True    True       True   
4             [71, 77): 'LONDON'      LOC   True    True       True   
..                           ...      ...    ...     ...        ...   
149  [590, 603): 'European tour'     MISC  False   False      False   
150        [962, 969): 'Campese'      LOC  False   False      False   
151          [39, 46): 'CAMPESE'     MISC  False   False      False   
152   [1333, 1343): 'Twickenham'      PER  False   False      False   
153      [516, 525): 'Wallabies'      LOC  False   False      False   

     carrerasb  chieu  curran  demeulder  florian  hammerton  hendrickx  \
0         True   True    True       True     True       True       True   
1         True   True   False      False    False      False      False   
2        False   True   False      False     True      False      False   
3         True   True    True       True     True      False       True   
4         True   True    True       True     True       True       True   
..         ...    ...     ...        ...      ...        ...        ...   
149      False  False   False      False    False      False      False   
150      False  False   False      False    False      False      False   
151      False  False   False      False    False      False      False   
152      False  False   False      False    False      False      False   
153      False  False   False      False    False      False      False   

     klein  mayfield  mccallum  munro  whitelaw     wu  zhang  
0     True      True      True   True      True   True   True  
1     True     False     False  False     False  False   True  
2    False     False     False  False     False  False  False  
3     True      True      True   True      True   True   True  
4     True      True      True   True      True   True   True  
..     ...       ...       ...    ...       ...    ...    ...  
149  False     False     False  False      True  False  False  
150  False     False     False  False      True  False  False  
151  False     False     False  False     False   True  False  
152  False     False     False  False     False   True  False  
153  False     False     False  False     False  False   True  

[154 rows x 19 columns]

In [11]:
# If you look at the above dataframe, you can see that some entities 
# ("RUGBY UNION", for example) are "easy", in that almost every entry
# found them correctly. Other entities, like "CAMPESE", are "harder",
# in that few of the entrants correctly identified them. Let's add
# a column that quantifies this "difficulty level" by counting how 
# many teams found each true or false positive.
for df in indicators:
    # Convert the teams' indicator columns into a single matrix of 
    # Boolean values.
    vectors = df[df.columns[3:]].values
    counts = np.count_nonzero(vectors, axis=1)
    df["num_teams"] = counts

# Show the dataframe for document 10 again, this time with the new
# "num_teams" column at the far right.
indicators[10]

token_span ent_type   gold  bender  carrerasa  \
0        [11, 22): 'RUGBY UNION'      ORG   True    True       True   
1             [24, 30): 'LITTLE'      PER   True    True       True   
2            [39, 46): 'CAMPESE'      PER   True   False      False   
3      [57, 70): 'Robert Kitson'      PER   True    True       True   
4             [71, 77): 'LONDON'      LOC   True    True       True   
..                           ...      ...    ...     ...        ...   
149  [590, 603): 'European tour'     MISC  False   False      False   
150        [962, 969): 'Campese'      LOC  False   False      False   
151          [39, 46): 'CAMPESE'     MISC  False   False      False   
152   [1333, 1343): 'Twickenham'      PER  False   False      False   
153      [516, 525): 'Wallabies'      LOC  False   False      False   

     carrerasb  chieu  curran  demeulder  florian  hammerton  hendrickx  \
0         True   True    True       True     True       True       True   
1         True   True   False      False    False      False      False   
2        False   True   False      False     True      False      False   
3         True   True    True       True     True      False       True   
4         True   True    True       True     True       True       True   
..         ...    ...     ...        ...      ...        ...        ...   
149      False  False   False      False    False      False      False   
150      False  False   False      False    False      False      False   
151      False  False   False      False    False      False      False   
152      False  False   False      False    False      False      False   
153      False  False   False      False    False      False      False   

     klein  mayfield  mccallum  munro  whitelaw     wu  zhang  num_teams  
0     True      True      True   True      True   True   True         16  
1     True     False     False  False     False  False   True          6  
2    False     False     False  False     False  False  False          2  
3     True      True      True   True      True   True   True         15  
4     True      True      True   True      True   True   True         16  
..     ...       ...       ...    ...       ...    ...    ...        ...  
149  False     False     False  False      True  False  False          1  
150  False     False     False  False      True  False  False          1  
151  False     False     False  False     False   True  False          1  
152  False     False     False  False     False   True  False          1  
153  False     False     False  False     False  False   True          1  

[154 rows x 20 columns]

In [12]:
# Now we can rank the entities in document 10 by "difficulty", either as 
# true positives for the models to find...
# (just for document 10 for the moment)
ind = indicators[10].copy()
ind[ind["gold"] == True].sort_values("num_teams").head(10)

token_span ent_type  gold  bender  carrerasa  carrerasb  \
2          [39, 46): 'CAMPESE'      PER  True   False      False      False   
21  [1020, 1030): 'Barbarians'      ORG  True   False      False      False   
38   [1687, 1696): 'All Black'      ORG  True   False      False      False   
10  [334, 346): 'Queenslander'     MISC  True   False      False      False   
34  [1536, 1546): 'Barbarians'      ORG  True   False       True      False   
7     [164, 174): 'Barbarians'      ORG  True    True      False      False   
28  [1333, 1343): 'Twickenham'      LOC  True    True      False      False   
1           [24, 30): 'LITTLE'      PER  True    True       True       True   
41  [1740, 1750): 'Barbarians'      ORG  True   False       True      False   
19     [761, 770): 'Wallabies'      ORG  True   False       True      False   

    chieu  curran  demeulder  florian  hammerton  hendrickx  klein  mayfield  \
2    True   False      False     True      False      False  False     False   
21  False   False      False    False      False      False   True     False   
38  False   False      False    False      False      False  False     False   
10   True   False       True     True      False      False  False     False   
34  False   False      False    False      False       True   True     False   
7   False   False       True    False      False       True  False     False   
28  False    True      False    False      False       True   True      True   
1    True   False      False    False      False      False   True     False   
41  False   False      False     True      False      False  False      True   
19   True    True      False     True      False       True  False     False   

    mccallum  munro  whitelaw     wu  zhang  num_teams  
2      False  False     False  False  False          2  
21      True  False     False  False  False          2  
38     False   True      True   True  False          3  
10     False  False     False  False   True          4  
34      True   True     False  False  False          5  
7       True   True     False  False  False          5  
28     False  False     False  False  False          5  
1      False  False     False  False   True          6  
41      True   True      True  False  False          6  
19      True   True     False   True  False          8

In [13]:
# ...or as false positives to avoid:
ind[ind["gold"] == False].sort_values("num_teams", ascending=False).head(10)

token_span ent_type   gold  bender  \
90                  [1020, 1030): 'Barbarians'     MISC  False    True   
91                  [1536, 1546): 'Barbarians'     MISC  False    True   
94                    [164, 174): 'Barbarians'     MISC  False   False   
104                         [24, 30): 'LITTLE'      LOC  False   False   
98                  [2013, 2023): 'Pontypridd'      ORG  False   False   
95   [334, 361): 'Queenslander Daniel Herbert'      PER  False   False   
96                  [1333, 1343): 'Twickenham'     MISC  False   False   
101                 [1333, 1343): 'Twickenham'      ORG  False   False   
102                  [1687, 1696): 'All Black'     MISC  False   False   
103                 [1740, 1750): 'Barbarians'     MISC  False   False   

     carrerasa  carrerasb  chieu  curran  demeulder  florian  hammerton  \
90        True       True   True    True      False     True      False   
91       False       True   True    True      False     True      False   
94        True      False   True    True      False     True      False   
104      False      False  False    True      False    False       True   
98        True       True  False   False      False    False      False   
95        True      False  False   False      False    False      False   
96        True      False  False   False       True     True      False   
101      False      False   True   False      False    False      False   
102      False      False   True    True      False     True      False   
103      False      False   True    True      False    False      False   

     hendrickx  klein  mayfield  mccallum  munro  whitelaw     wu  zhang  \
90        True  False      True     False  False     False   True   True   
91       False  False      True     False  False     False   True   True   
94       False   True      True     False  False     False   True   True   
104       True  False      True     False   True     False  False  False   
98        True   True     False      True  False     False  False  False   
95       False  False      True     False   True     False   True  False   
96       False  False     False     False  False     False  False   True   
101      False  False     False      True   True      True  False  False   
102      False  False     False     False  False     False  False   True   
103      False   True     False     False  False     False  False   True   

     num_teams  
90          10  
91           8  
94           8  
104          5  
98           5  
95           4  
96           4  
101          4  
102          4  
103          4

In [14]:
# To get a better picture of what entities are "difficult", we need to look 
# across the entire validation set. Let's combine the dataframes in 
# `indicators` into a single dataframe that covers all the documents.

# First we preprocess each dataframe to make it easier to combine.
to_stack = [
    pd.DataFrame({
        "doc_id": i,
        # TokenSpanArrays from different documents can't currently be stacked,
        # so convert to TokenSpan objects.
        "token_span" : indicators[i]["token_span"].astype(object),
        "ent_type": indicators[i]["ent_type"],
        "gold": indicators[i]["gold"],
        "num_teams": indicators[i]["num_teams"]
    })
    for i in range(len(indicators))
]

# Then we concatenate all the preprocessed dataframes into a single dataframe.
all_counts = pd.concat(to_stack)

all_counts

doc_id                             token_span ent_type   gold  num_teams
0        0                      [19, 24): 'JAPAN'      LOC   True         12
1        0                      [40, 45): 'CHINA'      PER   True          0
2        0                [66, 77): 'Nadim Ladki'      PER   True         15
3        0                     [78, 84): 'AL-AIN'      LOC   True         12
4        0      [86, 106): 'United Arab Emirates'      LOC   True         15
..     ...                                    ...      ...    ...        ...
50     230                 [19, 29): 'ENGLISHMAN'      LOC  False          1
51     230                 [428, 436): 'Charlton'      LOC  False          3
52     230  [1076, 1097): 'European championship'     MISC  False          1
53     230      [1346, 1363): 'World Cup winning'     MISC  False          1
54     230        [19, 38): 'ENGLISHMAN CHARLTON'      ORG  False          1

[11999 rows x 5 columns]

In [15]:
# Now we can pull out the most difficult entities across the entire validation
# set.
# First, let's find the most difficult entities from the standpoint of recall:
# entities that are in the gold standard, but not in most results.
difficult_recall = all_counts[all_counts["gold"] == True].sort_values("num_teams").reset_index(drop=True)
difficult_recall.head(10)

doc_id                                        token_span ent_type  gold  \
0     216                      [20, 36): 'SHEFFIELD SHIELD'     MISC  True   
1      54                        [3239, 3249): 'Full Light'     MISC  True   
2     177                              [11, 19): 'Honda RV'     MISC  True   
3      31                       [529, 542): '1. FC Cologne'      ORG  True   
4      54                            [1722, 1728): 'Okocim'      ORG  True   
5     149                  [1502, 1518): 'Consumer Project'      PER  True   
6      90                       [1129, 1140): 'Warsaw Pact'     MISC  True   
7      92  [536, 570): 'Movement for a Democratic Slovakia'      ORG  True   
8     216                            [308, 316): 'Victoria'      ORG  True   
9     216                            [179, 187): 'Victoria'      ORG  True   

   num_teams  
0          0  
1          0  
2          0  
3          0  
4          0  
5          0  
6          0  
7          0  
8          0  
9          0

In [16]:
# Hmm, everything is zero. How many entities were found by zero teams?  One team?
(all_counts[all_counts["gold"] == True][["num_teams", "token_span"]]
 .groupby("num_teams").count()
 .rename({"token_span": "count"}))
 # TODO: The last step here has no effect, due to a bug in Pandas. Fix the bug!

token_span
num_teams            
0                 140
1                  73
2                  88
3                  73
4                  99
5                  80
6                  85
7                  89
8                 125
9                 124
10                125
11                180
12                212
13                293
14                442
15               1040
16               2380

In [17]:
# Yikes! 140 entities in the validation set were so hard to find, they
# were extracted by 0 teams.
# Let's go back and look at some of those 0-team entities in context:
difficult_recall["context"] = difficult_recall["token_span"].apply(lambda t: t.context())
pd.set_option('max_colwidth', 100)
difficult_recall.head(20)

doc_id                                              token_span ent_type  \
0      216                            [20, 36): 'SHEFFIELD SHIELD'     MISC   
1       54                              [3239, 3249): 'Full Light'     MISC   
2      177                                    [11, 19): 'Honda RV'     MISC   
3       31                             [529, 542): '1. FC Cologne'      ORG   
4       54                                  [1722, 1728): 'Okocim'      ORG   
5      149                        [1502, 1518): 'Consumer Project'      PER   
6       90                             [1129, 1140): 'Warsaw Pact'     MISC   
7       92        [536, 570): 'Movement for a Democratic Slovakia'      ORG   
8      216                                  [308, 316): 'Victoria'      ORG   
9      216                                  [179, 187): 'Victoria'      ORG   
10     216                                  [166, 174): 'Tasmania'      ORG   
11      95                                   [1110, 1115): 'Simec'      ORG   
12     176                         [2777, 2792): 'PT Panutan Duta'      PER   
13     178  [1849, 1887): 'General Agreement on Tariffs and Trade'     MISC   
14     215                                    [42, 50): 'VICTORIA'      ORG   
15       8                               [237, 248): 'McDonald 's'      ORG   
16     146                                    [631, 637): 'ACCESS'     MISC   
17       8                                 [696, 705): 'Yorkshire'      LOC   
18      97                                   [125, 132): 'Quequen'      LOC   
19     146                              [143, 155): 'NYMEX ACCESS'     MISC   

    gold  num_teams  \
0   True          0   
1   True          0   
2   True          0   
3   True          0   
4   True          0   
5   True          0   
6   True          0   
7   True          0   
8   True          0   
9   True          0   
10  True          0   
11  True          0   
12  True          0   
13  True          0   
14  True          0   
15  True          0   
16  True          0   
17  True          0   
18  True          0   
19  True          0   

                                                                                                context  
0                                       [SHEFFIELD SHIELD] SCORE.\nHOBART, Australia 1996-12-07\nClo...  
1    ...centrating on its leading brand, Zywiec [Full Light], which accounts for 85 percent of sales...  
2                                               [Honda RV] exceeds sales target.\nTOKYO 1996-12-06\n...  
3   ...5 30 20 28\nVfL Bochum 16 7 6 3 23 21 27\n[1. FC Cologne] 16 8 2 6 31 27 26\nSchalke 04 17 7 ...  
4       ... while Carlsberg has the same amount in [Okocim].\nEarlier this year South African Brewer...  
5   ...r lobbyist heading the Washington-based [Consumer Project] on Technology.\n" None of the trea...  
6   ...which used to be part of the Soviet-led [Warsaw Pact], saying such moves would threaten its s...  
7   ...rime Minister Vladimir Meciar 's ruling [Movement for a Democratic Slovakia], was stripped of...  
8         ... 119, David Boon 118, Shaun Young 113); [Victoria] 220 for three( Dean Jones 130 not out).  
9     ...ield cricket match between Tasmania and [Victoria] at Bellerive Oval on Saturday:\nTasmania...  
10     ... Sheffield Shield cricket match between [Tasmania] and Victoria at Bellerive Oval on Satur...  
11       ...t significant volume among the gainers.\n[Simec], the steelmaking arm of the debt-ridden...  
12  ...find.\nBre-X has a partnership deal with [PT Panutan Duta] of the Panutan Group run by Presid...  
13  ...e commitments under its predecessor the [General Agreement on Tariffs and Trade]( GATT).\nIn ...  
14   ...OCSTART-\nCRICKET- JONES HITS CENTURY AS [VICTORIA] FIGHT BACK.\nHOBART, Australia 1996-12-0...  
15  ...taurant staff during a disturbance at a [McDonald 's] fast-food restaurant.\nBowyer, 19, who ...  
16      ...ged hands within the first few hours of [ACCESS

**Some of these entities are "difficult" because the validation set contains incorrect labels.**

For reference, there's a copy of the CoNLL labeling rules in this repository at
[resources/conll_03/ner/annotation.txt](../resources/conll_03/ner/annotation.txt)

There are 4 incorrect labels in this first set of 20:
* `[3289, 3299): 'Full Light'` should be "Zywiec Full Light"
* `[11, 19): 'Honda RV'` should be tagged `ORG`
* `[1525, 1541): 'Consumer Project'` should be "Consumer Project on Technology" and should be tagged `ORG`
* `[244, 255): 'McDonald 's'` should be tagged `MISC` (because it's an "adjective ... derived from a word which is ... organisation")

In [18]:
# Let's look at the entities that are difficult from the perspective of 
# precision: that is, in many models' results, but not in the gold standard.
difficult_precision = all_counts[all_counts["gold"] == False].sort_values("num_teams", ascending=False).reset_index(drop=True)

# Again, we can add some context to these spans:
difficult_precision["context"] = difficult_precision["token_span"].apply(lambda t: t.context())
difficult_precision.head(20)

doc_id                               token_span ent_type   gold  \
0      202                      [24, 31): 'BRITISH'     MISC  False   
1      207               [1305, 1315): 'Portsmouth'      ORG  False   
2      199                   [108, 116): 'Scottish'     MISC  False   
3      216                   [166, 174): 'Tasmania'      LOC  False   
4       40          [144, 161): 'Santiago Bernabeu'      LOC  False   
5      223               [232, 244): 'Philadelphia'      ORG  False   
6      216                   [308, 316): 'Victoria'      LOC  False   
7       36                  [349, 358): 'Karlsruhe'      ORG  False   
8      100                   [983, 991): 'Congress'      ORG  False   
9       36                   [398, 406): 'Freiburg'      ORG  False   
10      27           [714, 730): 'EASTERN DIVISION'     MISC  False   
11      70  [177, 202): 'New York Commodities Desk'      ORG  False   
12     223                   [256, 264): 'St Louis'      ORG  False   
13     170                    [627, 634): 'Bangkok'      LOC  False   
14      29                   [775, 783): 'Nebraska'      LOC  False   
15      38                [1116, 1125): 'Salamanca'      LOC  False   
16     155                    [777, 784): 'Antwerp'      LOC  False   
17     122                   [1319, 1325): 'German'     MISC  False   
18     112              [174, 187): 'John Mills Jr'      PER  False   
19     223                   [267, 275): 'COLORADO'      ORG  False   

    num_teams  \
0          16   
1          16   
2          16   
3          16   
4          16   
5          16   
6          16   
7          16   
8          16   
9          16   
10         16   
11         16   
12         16   
13         16   
14         16   
15         16   
16         16   
17         16   
18         16   
19         16   

                                                                                                context  
0                                                [BRITISH] RESULTS.\nLONDON 1996-12-07\nResults of B...  
1   ...2 26\nManchester City 22 8 2 12 26 35 26\n[Portsmouth] 22 7 5 10 25 29 26\nReading 22 7 5 10 ...  
2    ...W 1996-12-07\nLeading goalscorers in the\n[Scottish] premier division after Saturday 's matc...  
3      ... Sheffield Shield cricket match between [Tasmania] and Victoria at Bellerive Oval on Satur...  
4   ...ll breathalyse fans at the gates of the [Santiago Bernabeu] stadium and ban drunk supporters ...  
5   ...rgh 5 WASHINGTON 3\nMontreal 3 CHICAGO 1\n[Philadelphia] 6 DALLAS 3\nSt Louis 4 COLORADO 3\nE...  
6         ... 119, David Boon 118, Shaun Young 113); [Victoria] 220 for three( Dean Jones 130 not out).  
7   ...w 8th).\nHalftime 0-1.\nAttendance 33,000\n[Karlsruhe] 3( Reich 29th, Carl 44th, Dundee 69th)...  
8     ...n Congress would ratify the treaty with [Congress] quickly.\n'' The reactions from business...  
9   ... 3( Reich 29th, Carl 44th, Dundee 69th) [Freiburg] 0.\nHalftime 2-0.\nAttendance 33,000\nScha...  
10  ... 0 250 317\nNATIONAL FOOTBALL CONFERENCE\n[EASTERN DIVISION]\nW L T PF PA\nDALLAS 8 5 0 254 2...  
11              ...5.25 fio 35,000/ 30,000 China Steel.\n-- [New York Commodities Desk] +1 212 859 1640  
12            ...eal 3 CHICAGO 1\nPhiladelphia 6 DALLAS 3\n[St Louis] 4 COLORADO 3\nEDMONTON 5 Ottawa 2  
13                       ...0 Vietnam\nBut 2 27/11/96 5,000 Burma\n-- [Bangkok] newsroom( 662) 652-0642  
14    ...Roque and defensive end Jared Tomich of [Nebraska].\nThe Lombardi Award is presented to the...  
15   ...aulo Bento( Oviedo, Spain), Jose Taira( [Salamanca], Spain):\nForwards- Antonio Folha( Porto...  
16      ...kish man smuggled heroin from Turkey to [Antwerp] from where it was taken to Spain, Franc...  
17       ...incidents were chow chows, Rottweilers, [German] shepherds, cocker spaniels and Dalmatia...  
18  ...n whose trailer home he robbed in 1982, [John Mills Jr].\n, 41, was put to death in Florida '...  
19                       ...GO 1\n

**As with the entities in `difficult_recall`, some of these entities in `difficult_precision` are "difficult" because the validation set has missing and incorrect labels.**

**13** of these first 20 "incorrect" results are due to missing and incorrect labels:
* `[25, 32): 'BRITISH''` in document 202 should be tagged `MISC`.
* `[1317, 1327): 'Portsmouth'` in document 207 should be tagged `ORG`, not `LOC`.
* `[110, 118): 'Scottish'` in document 199 should be tagged `MISC`
  (or `[28, 53): 'SCOTTISH PREMIER DIVISION'` and 
  `[110, 135): 'Scottish premier division'` should both be tagged `ORG`).
* `[146, 163): 'Santiago Bernabeu'` in document 40 should be tagged `MISC`
  (because the "s" in `[146, 171): 'Santiago Bernabeu stadium'` is not capitalized).
* `[239, 251): 'Philadelphia'` in document 223 should be tagged `ORG`, not `LOC`.
* `[367, 376): 'Karlsruhe'` in document 36 should be tagged `ORG`, not `LOC`.
* `[1003, 1011): 'Congress'` in document 100 should be tagged `ORG`
  (also, `[957, 964): 'Chilean' ==> MISC` should be replaced with 
  `[957, 973): 'Chilean Congress' ==> ORG`).
* `[420, 428): 'Freiburg'` in document 36 should be tagged `ORG`, not `LOC`.
* In document 70, `[186, 211): 'New York Commodities Desk'`, not `[186, 206): 'New York Commodities'`, should be tagged `ORG`.
* `[263, 271): 'St Louis'` in document 223 should be tagged `ORG`, not `LOC`.
* `[788, 795): 'Antwerp'` in document 155 should be tagged `LOC`, not `ORG`.
* In document 112, `[178, 191): 'John Mills Jr'`, not `[178, 188): 'John Mills'`, should be tagged `PER`.
* `[274, 282): 'COLORADO'` in document 223 should be tagged `ORG`.


In [19]:
# Here's the gold standard data for document 155, for example.
# Note line 12.
gold_standard_spans[155][0:60]

token_span ent_type
0       [11, 18): 'Belgian'     MISC
1      [64, 72): 'BRUSSELS'      LOC
2       [170, 175): 'Spain'      LOC
3     [230, 237): 'Belgian'     MISC
4     [348, 355): 'Belgian'     MISC
5     [424, 431): 'Antwerp'      ORG
6     [460, 467): 'Belgian'     MISC
7   [538, 547): 'Barcelona'      LOC
8     [606, 613): 'Turkish'     MISC
9     [713, 720): 'Belgium'      LOC
10    [734, 741): 'Turkish'     MISC
11     [767, 773): 'Turkey'      LOC
12    [777, 784): 'Antwerp'      ORG
13      [812, 817): 'Spain'      LOC
14     [819, 825): 'France'      LOC
15    [830, 837): 'Germany'      LOC
16    [904, 911): 'Belgium'      LOC
17   [935, 943): 'European'     MISC
18  [1009, 1016): 'Belgian'     MISC

In [20]:
# The above gold standard spans in context. 
gold_standard_spans[155]["token_span"].values

<TokenSpanArray>
[    [11, 18): 'Belgian',    [64, 72): 'BRUSSELS',     [170, 175): 'Spain',
   [230, 237): 'Belgian',   [348, 355): 'Belgian',   [424, 431): 'Antwerp',
   [460, 467): 'Belgian', [538, 547): 'Barcelona',   [606, 613): 'Turkish',
   [713, 720): 'Belgium',   [734, 741): 'Turkish',    [767, 773): 'Turkey',
   [777, 784): 'Antwerp',     [812, 817): 'Spain',    [819, 825): 'France',
   [830, 837): 'Germany',   [904, 911): 'Belgium',  [935, 943): 'European',
 [1009, 1016): 'Belgian']
Length: 19, dtype: CharSpan